File: leetcode/131/solution.go
package main

var backtrack func(i int) [][]string

func partition(s string) [][]string {
	dp := make([][][]string, len(s))

	backtrack = func(i int) [][]string {
		arr := [][]string{}
		if len(s) <= i {
			// No more strings to be processed -> return an empty array
			return [][]string{{}}
		}
		if dp[i] != nil {
			return dp[i]
		}
		if len(s[i:]) == 1 {
			// Only one character -> return it
			subArr := make([]string, 1)
			subArr[0] = string(s[i])
			arr = append(arr, subArr)
			dp[i] = arr
			return dp[i]
		}

		for j := i + 1; j <= len(s); j++ {
			subString := s[i:j]
			if validPalindrome(subString) {
				for _, palindromes := range backtrack(j) {
					palindromes = append([]string{subString}, palindromes...)
					arr = append(arr, palindromes)
				}
			}
		}
		dp[i] = arr
		return dp[i]
	}

	return backtrack(0)
}

func validPalindrome(s string) bool {
	i, j := 0, len(s)-1
	for i < j {
		if s[i] != s[j] {
			return false
		}
		i++
		j--
	}
	return true
}

File: leetcode/131/solution.py
"""
https://leetcode.com/problems/palindrome-partitioning/
131. Palindrome Partitioning
Given a string s, partition s such that every substring of the partition is a palindrome. 
Return all possible palindrome partitioning of s.
"""
from typing import List, Tuple
from unittest import TestCase, main
from functools import lru_cache


"""
Backtrack + Dynamic Programming approach
"""


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def valid_palindrome(s: str) -> bool:
            return s != "" and s == s[::-1]

        @lru_cache()
        def backtrack(i: int) -> List[List[str]]:
            arr: List[List[str]] = []
            if len(s[i:]) == 0:
                # No more strings to be processed -> return empty array
                return [[]]

            if len(s[i:]) == 1:
                # Only one character -> return it
                return [[s[i]]]

            for j in range(i + 1, len(s) + 1):
                substring = s[i:j]
                if valid_palindrome(substring):
                    for palindromes in backtrack(j):
                        arr.append([substring] + palindromes)

            return arr

        return backtrack(0)


class Test(TestCase):
    data: List[Tuple[str, List[List[str]]]] = [
        ("bb", [["b", "b"], ["bb"]]),
        ("aab", [["a", "a", "b"], ["aa", "b"]]),
        ("a", [["a"]]),
    ]

    def test_solution(self):
        solution = Solution()
        for s, expected in self.data:
            self.assertEqual(solution.partition(s), expected)


if __name__ == "__main__":
    main()
